File: Data Structures & Algorithms/is-subsequence/submission-1.py
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if not s:
            return True

        m, n = len(t), len(s)

        if m < n:
            return False

        i = 0

        for c in t:
            if c == s[i]:
                i += 1

                if i >= n:
                    return True

        return False

File: Data Structures & Algorithms/palindrome-partitioning/submission-1.cpp
class Solution {
    string s;
    vector<string> part;
    vector<vector<string>> res;

    void dfs(int i) {
        if (i == s.length()) {
            res.push_back(part);
            return;
        }

        for (int j = i; j < s.size(); ++j) {
            if (is_pali(i, j)) {
                part.push_back(s.substr(i, j - i + 1));
                dfs(j + 1);
                part.pop_back();
            }
        }
    }

    bool is_pali(int l, int r) {
        while (l < r) {
            if (s[l] != s[r]) {
                return false;
            }
            ++l, --r;
        }
        return true;
    }

public:
    vector<vector<string>> partition(string s) {
        this->s = s;
        dfs(0);
        return res;
    }
};
